### PR TITLE
Adds contract verification via sourcify

### PR DESCRIPTION
**Pull request by robertabbott**
## Motivation
Fixes https://github.com/foundry-rs/foundry/issues/1822

## Solution
Adds an additional option `--verification-provider` that allows the user to specify and contract-verifier other than etherscan. Omitting the option causes verify to default to etherscan. If `--verification-provider sourcify` is passed to `forge create` the contract is verified by hitting the[ sourcify api](https://docs.sourcify.dev/docs/api/server/check-by-addresses/)

### CODE REVIEW

File: cli/src/cmd/forge/verify.rs [hunk 1]
```diff
@@ -177,6 +228,41 @@ impl VerifyArgs {
         Ok(())
     }
 
+    async fn verify_with_sourcify(self) -> eyre::Result<()> {
+        let url = format!(
+            "https://sourcify.dev/server/check-by-addresses?addresses={}&chainIds={}",
+            hex::encode(self.address.as_bytes()),
```
Comment (mattsse) [inline cli/src/cmd/forge/verify.rs:234]: this formats with 0x prefix, note debug formatting is necessary here
```suggestion
            format!("{:?}", self.address),
```

File: cli/src/cmd/forge/verify.rs [hunk 2]
```diff
@@ -177,6 +228,41 @@ impl VerifyArgs {
...
+            self.chain.id(),
+        );
+        let response = get(url).await?;
+        if !response.status().is_success() {
+            eprintln!(
+                "Sourcify verification request for address ({}) failed with status code {}",
+                hex::encode(self.address.as_bytes()),
```
Comment (mattsse) [inline cli/src/cmd/forge/verify.rs:241]: same

File: cli/src/cmd/forge/verify.rs [hunk 3]
```diff
@@ -177,6 +228,41 @@ impl VerifyArgs {
...
+                response.status()
+            );
+            std::process::exit(1);
+        };
+
+        // Response looks like:
+        // `[{"address":"0x68B1D87F95878fE05B998F19b66F4baba5De1aed","status":"false"}]`
+        let parsed_response = response.json::<serde_json::Value>().await?;
```
Comment (mattsse) [inline cli/src/cmd/forge/verify.rs:249]: can we add proper bindings for this?
https://app.quicktype.io/ generates rust types for json

File: cli/src/cmd/forge/verify.rs [hunk 4]
```diff
@@ -23,18 +23,54 @@ use foundry_utils::Retry;
...
+                write!(f, "Etherscan")?;
+            }
+            VerificationProvider::Sourcify => {
+                write!(f, "Sourcify")?;
+            }
+        };
+        Ok(())
+    }
+}
+
```
Comment (mattsse) [inline cli/src/cmd/forge/verify.rs:73]: let's move these below the `VerifyArgs`, so each cmd file starts with the actual command

**Review by mattsse — CHANGES_REQUESTED** [review]
nice!

only a few of smol nits,

a basic integration test would be nice, we have some etherscan verify integration tests here:
https://github.com/foundry-rs/foundry/blob/master/cli/tests/it/verify.rs

the `forgetest!` macro creates a new test and verify tests are currently run if the necessary env vars are set

**Comment by robertabbott** [comment]
re: https://github.com/foundry-rs/foundry/pull/1872#discussion_r892217725

there is one result / array entry per address in the request, so there will only ever be a single item

File: cli/tests/it/verify.rs [hunk 5]
```diff
@@ -148,14 +151,20 @@ fn verify_flag_on_create_on_chain(
 ) {
     // only execute if keys present
     if let Some(info) = info {
-        add_unique(&prj);
-        add_verify_target(&prj);
+        for verifier in VERIFICATION_PROVIDERS {
```
Comment (mattsse) [inline cli/tests/it/verify.rs:154]: nice!

**Review by mattsse — APPROVED** [review]
lgtm

File: cli/src/cmd/forge/verify.rs [hunk 6]
```diff
@@ -109,11 +113,26 @@ pub struct VerifyArgs {
         value_name = "LIBRARIES"
     )]
     pub libraries: Vec<String>,
+
+    #[clap(
+        long = "verifier",
+        help_heading = "VERIFICATION PROVIDER",
+        help = "Contract verification provider to use `sourcify` or `etherscan` [Default: etherscan]",
```
Comment (onbjerg) [inline cli/src/cmd/forge/verify.rs:120]: Same as above

```suggestion
        help = "Contract verification provider to use `sourcify` or `etherscan`",
```

**Review by onbjerg — APPROVED** [review]
Two small nits otherwise LGTM

**Comment by iainnash** [comment]
Would it be possible to have an option to verify with multiple providers? or would the idea be to run the command twice in that case?

File: cli/src/cmd/forge/verify.rs [hunk 7]
```diff
@@ -177,6 +196,34 @@ impl VerifyArgs {
         Ok(())
     }
 
+    async fn verify_with_sourcify(self) -> eyre::Result<()> {
+        let url = format!(
+            "https://sourcify.dev/server/check-by-addresses?addresses={}&chainIds={}",
```
Comment (kuzdogan) [inline cli/src/cmd/forge/verify.rs:201]: Not sure if I get it right but this is hitting the "Lookup" endpoint not the verification. `check-by-addresses` looks if the contract is already verified on Sourcify. For verification check out [`/verify`](https://docs.sourcify.dev/docs/api/server/v1/verify/)

**Comment by kuzdogan** [comment]
👀 Following 👀 

Let me know if I can help

**Comment by tynes** [comment]
Definitely interested in building on this PR as an upcoming Optimism testnet will be using blockscout as a block explorer and they have a contract verification API, it would be nice to also be able to support that. See https://github.com/EricxStone/hardhat-blockscout-verify

**Comment by tynes** [comment]
Just wondering the status on this PR, would be great to have this merged :)

**Comment by gakonst** [comment]
currently just checks for the status of the verification, but doesn't submit it! need someone to take it over the line

**Comment by kuzdogan** [comment]
Not Foundry but Optimism Goerli support is added to Sourcify so you can now verify contracts through the API and other tooling we have https://github.com/ethereum/sourcify/pull/773/files @tynes

**Comment by rkrasiuk** [comment]
@gakonst fyi taking over

**Comment by rkrasiuk** [comment]
@tynes current etherscan verifier is fully compatible with blockscout. see comments to https://github.com/gakonst/ethers-rs/pull/1343

**Comment by rkrasiuk** [comment]
closing in favor of #2748

**Pull request closed without merging**